### DISCUSSION TITLE
Clever Ways for Zero Downtime?

### DISCUSSION THREAD

**Discussion by dm17** [nginx-proxy/nginx-proxy#2431] [Q&A]
How do you guys upgrade your services running in production? Say you have new images for a bunch of services contained in docker-compose. If I want to run the latest versions, then I need to do docker-compose pull && docker-compose down && docker-compose up. At the very least it would be nice to have a "we'll be back in a few seconds" message instead of an nginx error that gets displayed...
Would you mind sharing how you all do this?
Thanks!

**Comment by hasnat** [comment]
Either have main nginx proxy config changed to use custom error templates. Or keep one container set as default host, which only serves "we'll be back in a few seconds" message. So in event of any container down that acts as backup server.
You don't need docker-compose down, just docker-compose up would replace containers to be upgraded.

**Comment by dm17** [comment]
@hasnat Much appreciated. Can you help me understand how 'docker-compose up' would replace containers to be upgraded? So if I build new containers, commit them to a private repo, do a 'docker-compose pull'... So how could I have the old containers running until the new containers are up (and prevent them from interfering during this process)? I don't think just doing 'docker-compose up' will do that for me, will it?
Thanks again @hasnat

**Comment by dm17** [comment]
I'm looking this over now... I hear it can do what I described above:
https://docs.docker.com/engine/swarm/stack-deploy/

**Comment by hasnat** [comment]
@dm17 for containers where image ins't changed or any of external envs being given don't change docker-compose up would be noop. you'd see messages like container up to date instead of recreating

**Comment by dm17** [comment]
I asked some docker experts and several of them answered that the way to do this is docker-swam. More specifically, to handle "zero downtime service upgrades."

**Comment by FedericoPonzi** [comment]
If you're using docker-compose, then this might help: https://docs.docker.com/compose/production/#deploying-changes

**Comment by dm17** [comment]
> If you're using docker-compose, then this might help: https://docs.docker.com/compose/production/#deploying-changes

Thanks... Typical of docker to assume so much in its documentation. They described it in once sentence: "This first rebuilds the image for web and then stop, destroy, and recreate just the web service." Therefore, since it does 'stop' first, it will be down until the new container is 'up', right? So this isn't as good as using service discovery - because if the new container takes 30s to start, then the downtime is 30s. Same as doing "docker-compose down && docker-compose up" after changes are made to the images, it seems.

**Comment by FedericoPonzi** [comment]
I haven't tried this, but one thing you could try is [blue-green deployment](https://martinfowler.com/bliki/BlueGreenDeployment.html):
* Deploy a new version of your compose with another service (`a-1.1`) which runs the new version of your service (`a-1.0`). You load balance between the old instance and the new.
* Wait for it to start
* Deploy a docker-compose where you remove the old instance to the balancing (`a-1.0`)
* Deploy a final new docker-compose version, where you have removed the `a-1.0` and you just have the new service (`a-1.1`).

**Comment by dm17** [comment]
> I haven't tried this, but one thing you could try is [blue-green deployment](https://martinfowler.com/bliki/BlueGreenDeployment.html):
> 
> * Deploy a new version of your compose with another service (`a-1.1`) which runs the new version of your service (`a-1.0`). You load balance between the old instance and the new.
> * Wait for it to start
> * Deploy a docker-compose where you remove the old instance to the balancing (`a-1.0`)
> * Deploy a final new docker-compose version, where you have removed the `a-1.0` and you just have the new service (`a-1.1`).

Thanks, I hadn't heard of this... It'll take some looking into, so first I'm gonna look into Traefik, which supposedly can do this via "service discovery" (I would be moving from nginx).
